kernel/thread.c

```diff
@@ -348,40 +410,172 @@ long myst_syscall_wait4(
 
-            if (match)
+            if (_wait_matcher("zombie", pid, process, p))
             {
                 if (wstatus)
-                    *wstatus = (p->exit_status << 8);
-
+                {
+                    *wstatus = 0;
+
+                    if (p->terminating_signum)
+                    {
+                        *wstatus = (p->terminating_signum & 0x7f);
+                    }
+                    else
+                    {
+                        *wstatus = ((p->exit_status & 0xff) << 8);
+                    }
+#ifdef TRACE
+                    printf("*** WSTATUS DETAILS: %d\n", *wstatus);
+                    printf(
+                        "    WIFEXITED (exited through exit(), _exit() or "
+                        "return from main): %s\n",
+                        WIFEXITED(*wstatus) ? "true" : "false");
+                    printf(
+                        "    WEXITSTATUS (return code via exit(), _exit() or "
+                        "return from main): %u\n",
+                        WEXITSTATUS(*wstatus));
+                    printf(
+                        "    WIFSIGNALED (exited via signal): %s\n",
+                        WIFSIGNALED(*wstatus) ? "true" : "false");
+                    printf(
+                        "    WTERMSIG (signal code to terminate): %u\n",
+                        WTERMSIG(*wstatus));
+                    printf(
+                        "    WCOREDUMP (signaled and dumped core, not "
+                        "supported): %s\n",
+                        WCOREDUMP(*wstatus) ? "true" : "false");
+                    printf(
+                        "    WIFSTOPPED (signaled to stop, not supported): "
+                        "%s\n",
+                        WIFSTOPPED(*wstatus) ? "true" : "false");
+                    if (WIFSTOPPED(*wstatus))
+                        printf(
+                            "    WSTOPSIG (signal code to stop, not "
+                            "supported): "
+                            "%u\n",
+                            WSTOPSIG(*wstatus));
+                    printf(
+                        "    WIFCONTINUED (signaled to resumed after stop, not "
+                        "supported): %s\n",
+                        WIFCONTINUED(*wstatus) ? "true" : "false");
+#endif
+                }
                 ret = p->pid;
+
+                // remove from zombie list
+                if (p->zprev == NULL)
+                {
+                    _zombies_head = p->znext;
+                }
+                else
+                {
+                    p->zprev->znext = p->znext;
+                }
+                if (p->znext != NULL)
+                    p->znext->zprev = p->zprev;
+
+                if (_zombies_tail == p)
+                    _zombies_tail = p->group_prev;
+
+                _zombies_count--;
+
+                // remove from process list
+                myst_spin_lock(&myst_process_list_lock);
```

Comment:
Possible dead lock due to acquiring a second lock while holding `_zombies_mutex`. Shouldn't the process be removed from the process list during zombify?

Reply:
The impact at this point is unknown for changing the lists threads are in. Moving to use the same lock for zombies and processes. Moving an item from one list to the other with two locks has the possibility of another thread thinking the process thread does not exist.